Add support for native histograms to fmtutil

mem opened the pull request:
util/fmtutil supports creating a write request including histograms, but it doesn't support native histograms. This change adds that missing support.

Note that `promtool` uses `fmtutil.MetricTextToWriteRequest`, which is affected by this change. The immediate intention of this change _is not_ to get `promtool` to support native histograms, but to be able to take a set of metric families as produced by e.g. `registry.Gather` and convert that into a write request.

mem commented:
I'm still trying to figure out how to test this change.

beorn7 reviewed (commented):
Thanks for doing this.

Note that `promtool` is currently only accepting metric files in the classic Prometheus text format, which doesn't support native histograms (and probably never will, assuming that native histogram support will be added to OpenMetrics or its successor). For a short term solution, maybe allow promtool to read protobuf (in binary or text form).

mem commented:
> Thanks for doing this.
> 
> Note that `promtool` is currently only accepting metric files in the classic Prometheus text format, which doesn't support native histograms (and probably never will, assuming that native histogram support will be added to OpenMetrics or its successor). For a short term solution, maybe allow promtool to read protobuf (in binary or text form).

Thanks for taking a look, that actually helps a lot.

I'll take another look.

mem commented on util/fmtutil/format.go:
```diff
@@ -133,25 +133,21 @@ func makeTimeseries(wr *prompb.WriteRequest, labels map[string]string, m *dto.Me
-			bucketLabels[model.BucketLabel] = fmt.Sprint(b.GetUpperBound())
-			toTimeseries(wr, bucketLabels, timestamp, float64(b.GetCumulativeCount()))
+
+		// TODO(mem): should this be else? Is a classic histogram without buckets even possible?
```
That makes sense, I think.

For a classic histogram, if I don't have buckets, I'm left with the sample count and the sample sum. Looking at the Go code, I can see that the count and the sum will be populated regardless of whether the histogram is native or classic.

With the Go client, if I _try_ to create a histogram without buckets (by specifying an empty slice), I still get the default buckets. The only way (with that current code) that I can find to make it respect the empty bucket list is to make it a native histogram (specify the native histogram bucket factor).

So now the part where I'm confused: I think you are right, a classic histogram without buckets seems to be possible (just by looking at the code), and it will have only the count and sum (and possibly the +inf bucket, by definition?). This does not seem to match the implementation.

mem replied:
And just to clarify what I mean by the last paragraph: the code I looked at seems to be fine with an empty bucket list. I cannot spot any assumption that the bucket list is populated. So I _think_ that if the bucket list was in fact empty, the code would work.

beorn7 replied:
> Looking at the Go code, I can see that the count and the sum will be populated regardless of whether the histogram is native or classic.

But in case of a native histogram, the count and sum is part of the complex histogram sample, while in the case of the classic histogram, it will spawn to new series (with new names due to "magic" suffixing).

> With the Go client, if I try to create a histogram without buckets (by specifying an empty slice), I still get the default buckets.

IIRC correctly, the intention here was to make users use a Summary without quantiles instead of a Histogram without buckets. But creative users will come up with ways of creating a bucketless classic histogram, I'm sure…
I guess it _should_ have the +Inf bucket then, but if it doesn't, Prometheus will probably tolerate it.

OpenMetrics will have ruled those edge cases more strictly, but I think our code should still be tolerant if a histogram comes in that doesn't comply strictly with the standard, but we can still make sense of it.

mem commented:
@beorn7 I modified some of the code (mostly added comments), copied the function you mentioned (I didn't want to export it from the other package) and added a couple of test cases.

The testcases are basically the actual output of this function. For the classic histogram case I can verify that it's correct just by visually inspecting it.

For the native histogram I'm not sure how to verify that the spans and offsets are correct, but I'm going to blindly trust that they are. The question is whether those are the values that should be in those fields.

Also, I need to figure out how to generate a native histogram with float counts.

mem commented on util/fmtutil/format_test.go:
```diff
@@ -80,17 +82,17 @@ var writeRequestFixture = &prompb.WriteRequest{
+				{Name: "__name__", Value: "http_request_duration_seconds_count"},
 				{Name: "job", Value: "promtool"},
 			},
-			Samples: []prompb.Sample{{Value: 53423, Timestamp: 1}},
+			Samples: []prompb.Sample{{Value: 144320, Timestamp: 1}},
 		},
 		{
 			Labels: []prompb.Label{
-				{Name: "__name__", Value: "http_request_duration_seconds_count"},
+				{Name: "__name__", Value: "http_request_duration_seconds_sum"},
 				{Name: "job", Value: "promtool"},
 			},
-			Samples: []prompb.Sample{{Value: 144320, Timestamp: 1}},
+			Samples: []prompb.Sample{{Value: 53423, Timestamp: 1}},
```
I'm not 100% sure why this is changing.

The output is supposed to be ordered by metric name, so bucket comes before count before sum. The order _should_ be the order shown here, and I don't think I'm actually changing something to modify that...

beorn7 replied:
The old code simply used the "wrong" order. (I don't think that the remote-write spec requires this kind of ordering, but it cannot harm to do it your way.) tl;dr: This is fine.

fgouteroux replied:
Yes there is no remote write spec about this ordering, I was just following this rule : label names sorted in lexicographical order.

mem commented on util/fmtutil/format.go:
```diff
@@ -201,3 +199,102 @@ func makeLabelsMap(m *dto.Metric, metricName string, extraLabels map[string]stri
+	})
+}
+
+// isNativeHistogram returns false iff the provided histograms has no spans at
+// all (neither positive nor negative) and a zero threshold of 0 and a zero
+// count of 0. In principle, this could still be meant to be a native histogram
+// with a zero threshold of 0 and no observations yet. In that case,
+// instrumentation libraries should add a "no-op" span (e.g. length zero, offset
+// zero) to signal that the histogram is meant to be parsed as a native
+// histogram. Failing to do so will cause Prometheus to parse it as a classic
+// histogram as long as no observations have happened.
+func isNativeHistogram(h *dto.Histogram) bool {
+	return len(h.GetPositiveSpan()) > 0 ||
+		len(h.GetNegativeSpan()) > 0 ||
+		h.GetZeroThreshold() > 0 ||
+		h.GetZeroCount() > 0
+}
```
I have been debating over how to handle this.

This function is a copy of the one found in [model/textparse/protobufparse.go](https://github.com/prometheus/prometheus/blob/83891135c6585fc2e9990824c9d12478acfe6b10/model/textparse/protobufparse.go#L575)

I would not mind the duplicate except that the chance of the original changing over time is non-zero.

One way is to play linker games. It's straightforward, but people might not like it.

The other is to move parts of `model/textparse` to an internal package (e.g. internal/model/textparse) and export this function from there so that this code can import it.

The other is to _export_ the function, sealing the contract regarding its future availability.

Any preference?

beorn7 replied:
tl;dr: I think we should stick with copying for now.

The problem is that the `fmtutil` version of this function (this one) imports `dto` as `dto "github.com/prometheus/client_model/go"`. The version in `textparse` umports `dto` as `dto "github.com/prometheus/prometheus/prompb/io/prometheus/client"`. The former is proto2 based and using the official Go protobuf library, the latter is proto3 based and using gogo protobuf. There is no difference on the wire, but the Go types are different.

The whole situation is very messy, but cleaning that up is not easy (and out of scope of this PR).

mem commented:
@fgouteroux could you please take a look at this, since you added the original code? Thanks!

beorn7 commented on util/fmtutil/format.go:
```diff
@@ -133,25 +133,23 @@ func makeTimeseries(wr *prompb.WriteRequest, labels map[string]string, m *dto.Me
+		//
+		// Note that this comes _after_ native histogams because calling
+		// addClassicHistogramToWriteRequest modifies the labels map.
+		if len(h.Bucket) > 0 {
```
I think this should be `if len(h.Bucket) > 0 || !isNativeHistogram(h) {`

Rationale: If there are classic buckets, then yes, you want this to be represented as a classic histogram, maybe even in parallel with the native histogram.

However, even if there are no buckets, this might be the edge case explained in your doc comment above. We can only say for sure if it is _not_ a native histogram. So in that case we should actually create that (arguably incomplete) classic histogram.

beorn7 commented on util/fmtutil/format.go:
```diff
@@ -201,3 +199,102 @@ func makeLabelsMap(m *dto.Metric, metricName string, extraLabels map[string]stri
+	metricName := labels[model.MetricNameLabel]
+
+	// Preserve metric name order with first bucket suffix timeseries then
+	// count suffix timeseries and finally sum suffix timeserie
```
Missing `s.` at the end (note the period).

beorn7 commented on util/fmtutil/format.go:
```diff
@@ -201,3 +199,102 @@ func makeLabelsMap(m *dto.Metric, metricName string, extraLabels map[string]stri
+		PositiveDeltas: h.GetPositiveDelta(),
+		PositiveCounts: h.GetPositiveCount(),
+		Timestamp:      timestamp,
+		// ResetHint: what is this?
```
If the remote-write request is created from data in the TSDB, you can populate the ResetHint. But if you just have a single dto protobuf, you have to leave it as "unknown" (which is the default, I hope).

beorn7 commented on util/fmtutil/format.go:
```diff
@@ -201,3 +199,102 @@ func makeLabelsMap(m *dto.Metric, metricName string, extraLabels map[string]stri
+	}
+
+	if h.SampleCountFloat != nil {
+		// Assume float histogram
```
Nit: End with period.

beorn7 reviewed (changes requested):
Looks good overall. Just some small comments left.

fgouteroux commented:
> @fgouteroux could you please take a look at this, since you added the original code? Thanks!

Oh it's a nice enhancement. I will be not very helpful there because I never used native histogram. Thanks to @beorn7 for the explainations.

I just saw this: 

> After years of research by Björn Rabenstein, we have designed a new native histogram that is supported starting with Prometheus v2.40.0.

So this PR is reviewed by the most skilled guy in native histogram 😉

beorn7 commented:
@mem fine then, I guess you only need to address my minor comments above, and we are good to go.

beorn7 commented:
@mem are you still on this? It feels there is very little left to make this ready for merge.

beorn7 commented:
@mem another friendly :bellhop_bell: . Would be great if you could invest the little bit of effort to get this past the finish line.

beorn7 commented:
In an effort to get our review backlog under control, I'll close this PR as stale. @mem that should in now way discourage you from picking up the work here. Simply reopen it once you are doing so.
